File: LeetCode/16 M5 Leetcode 42 - Trapping Rain Water/2pointers/42external.js
r
// https://leetcode.com/problems/trapping-rain-water/discuss/312091/Javascript-2-pointer

/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    if (height.length === 0) return 0;
    
    let i = 0;
    let j = height.length-1;
    
    let leftMax = 0;
    let rightMax = 0;
    let result = 0;
    
    while(i<=j){
        if (rightMax > leftMax){
            if (height[i] > leftMax) leftMax = height[i];
            else result += leftMax-height[i];
            i++;
        }else{
            if (height[j] > rightMax) rightMax = height[j];
            else result += rightMax-height[j];
            j--;
        }
    }
    return result;
};

File: LeetCode/16 M5 Leetcode 42 - Trapping Rain Water/2pointers/42-external.py
# https://leetcode.com/problems/trapping-rain-water/discuss/759078/Python-solution

class Solution(object):
    def trap(self, height):
        n = len(height)
        if n < 3:
            return 0

        result, left, right = 0, 1, n-2
        left_max = height[0]
        right_max = height[n-1]

        while left <= right:
            if left_max <= right_max:
                left_max = max(left_max, height[left])
                result += left_max - height[left]
                left += 1
            else:
                right_max = max(right_max, height[right])
                result += right_max - height[right]
                right -= 1

        return result
